File: clients/esp32/components/transport/test/test_offline_queue.py
"""
Host-side verification of the ESP32 transport offline chat queue (R5d).

Byte-for-byte Python port of the ring-buffer helpers in transport.c. Runs
with just Python 3 — no C toolchain required. If the C changes, update
both this file and test_offline_queue.c.

Run:
    python3 test_offline_queue.py
"""

import sys

# Constants — mirror transport.c
OFFLINE_QUEUE_CAP = 8
OFFLINE_QUEUE_MAX_TEXT = 1024
OFFLINE_QUEUE_MAX_SESSION = 64
OFFLINE_QUEUE_MAX_MSGID = 40


class OfflineQueue:
    """Python port of the s_oq ring buffer + enqueue/peek/clear semantics."""

    def __init__(self):
        self._buf = [None] * OFFLINE_QUEUE_CAP
        self._head = 0
        self._count = 0

    def enqueue(self, scope, text, client_msg_id):
        if text is None or client_msg_id is None:
            return False
        # Null-terminator semantics: strlen >= max → reject.
        if len(text) >= OFFLINE_QUEUE_MAX_TEXT:
            return False
        if len(client_msg_id) >= OFFLINE_QUEUE_MAX_MSGID:
            return False
        if scope is not None and len(scope) >= OFFLINE_QUEUE_MAX_SESSION:
            return False

        if self._count == OFFLINE_QUEUE_CAP:
            # Drop oldest.
            self._head = (self._head + 1) % OFFLINE_QUEUE_CAP
            self._count -= 1

        slot = (self._head + self._count) % OFFLINE_QUEUE_CAP
        self._buf[slot] = {
            "scope": scope if scope is not None else "",
            "text": text,
            "client_msg_id": client_msg_id,
        }
        self._count += 1
        return True

    def size(self):
        return self._count

    def peek(self, idx):
        if idx >= self._count:
            return None
        slot = (self._head + idx) % OFFLINE_QUEUE_CAP
        return self._buf[slot]

    def clear(self):
        self._head = 0
        self._count = 0

    def pop_head(self):
        """Test-only: simulate successful flush of one entry."""
        if self._count == 0:
            return False
        self._head = (self._head + 1) % OFFLINE_QUEUE_CAP
        self._count -= 1
        return True


# ---- Tests -----------------------------------------------------------------

failures = []


def check(cond, msg):
    if not cond:
        failures.append(msg)


def test_enqueue_three_fifo_order():
    q = OfflineQueue()
    check(q.size() == 0, "fresh size=0")
    check(q.enqueue("s1", "hello", "m1"), "enq 1")
    check(q.enqueue("s1", "world", "m2"), "enq 2")
    check(q.enqueue("s2", "goodbye", "m3"), "enq 3")
    check(q.size() == 3, "size=3")

    e0 = q.peek(0)
    check(
        e0["scope"] == "s1" and e0["text"] == "hello" and e0["client_msg_id"] == "m1",
        "FIFO idx=0 oldest",
    )
    e1 = q.peek(1)
    check(e1["text"] == "world", "FIFO idx=1")
    e2 = q.peek(2)
    check(e2["text"] == "goodbye", "FIFO idx=2 newest")
    check(q.peek(3) is None, "peek past end is None")


def test_overflow_drops_oldest():
    q = OfflineQueue()
    for i in range(10):
        q.enqueue("s", f"t{i}", f"m{i}")
    check(q.size() == OFFLINE_QUEUE_CAP, "cap holds at 8")
    check(q.peek(0)["client_msg_id"] == "m2", "head=m2 (m0/m1 evicted)")
    check(q.peek(OFFLINE_QUEUE_CAP - 1)["client_msg_id"] == "m9", "tail=m9")


def test_clear():
    q = OfflineQueue()
    q.enqueue("s", "a", "m1")
    q.enqueue("s", "b", "m2")
    check(q.size() == 2, "size=2 before clear")
    q.clear()
    check(q.size() == 0, "size=0 after clear")
    check(q.peek(0) is None, "cleared peek rejected")


def test_text_cap_rejects():
    q = OfflineQueue()
    big = "x" * (OFFLINE_QUEUE_MAX_TEXT + 16)
    check(not q.enqueue("s", big, "m1"), "oversized text rejected")
    check(q.size() == 0, "size unchanged after reject")

    at_cap = "y" * (OFFLINE_QUEUE_MAX_TEXT - 1)
    check(q.enqueue("s", at_cap, "m1"), "at-cap text accepted")
    check(q.size() == 1, "accepted increments size")


def test_null_args_rejected():
    q = OfflineQueue()
    check(not q.enqueue("s", None, "m1"), "None text rejected")
    check(not q.enqueue("s", "hi", None), "None msg_id rejected")
    check(q.size() == 0, "no enqueue on None")


def test_drain_preserves_order():
    q = OfflineQueue()
    q.enqueue("s", "first", "m1")
    q.enqueue("s", "second", "m2")
    q.enqueue("s", "third", "m3")
    check(q.peek(0)["text"] == "first", "drain: first")
    q.pop_head()
    check(q.peek(0)["text"] == "second", "drain: second")
    q.pop_head()
    check(q.peek(0)["text"] == "third", "drain: third")
    q.pop_head()
    check(q.size() == 0, "drain: empty")


def test_wraparound():
    q = OfflineQueue()
    for i in range(OFFLINE_QUEUE_CAP):
        q.enqueue("s", f"t{i}", f"m{i}")
    q.pop_head()
    q.pop_head()
    check(q.size() == 6, "size=6 after drain 2")
    q.enqueue("s", "new1", "mn1")
    q.enqueue("s", "new2", "mn2")
    check(q.size() == 8, "refilled to cap")
    check(q.peek(0)["client_msg_id"] == "m2", "wraparound head=m2")
    check(q.peek(7)["client_msg_id"] == "mn2", "wraparound tail=mn2")


if __name__ == "__main__":
    test_enqueue_three_fifo_order()
    test_overflow_drops_oldest()
    test_clear()
    test_text_cap_rejects()
    test_null_args_rejected()
    test_drain_preserves_order()
    test_wraparound()

    if failures:
        print(f"FAIL: {len(failures)} assertion(s)")
        for f in failures:
            print(f"  - {f}")
        sys.exit(1)
    print("All tests passed.")

File: clients/esp32/components/transport/test/test_offline_queue.c
/*
 * test_offline_queue.c — host-side unit test for the pure RAM FIFO
 * used by the ESP32 transport's chat offline queue (R5d).
 *
 * Build & run without ESP-IDF:
 *
 *   cc test_offline_queue.c -o test_offline_queue
 *   ./test_offline_queue
 *
 * The body below is a byte-for-byte copy of the enqueue/peek/clear/size
 * functions in transport.c — with ESP_LOGW reduced to fprintf. If you
 * change one, change both.
 */

#include <assert.h>
#include <stdio.h>
#include <stdint.h>
#include <stdbool.h>
#include <string.h>

/* Mirror constants from transport.c / transport_limits.h */
#define MOUMANTAI_MAX_ID_LEN 64
#define OFFLINE_QUEUE_CAP 8
#define OFFLINE_QUEUE_MAX_TEXT 1024
#define OFFLINE_QUEUE_MAX_SESSION MOUMANTAI_MAX_ID_LEN
#define OFFLINE_QUEUE_MAX_MSGID 40

typedef struct {
    char scope[OFFLINE_QUEUE_MAX_SESSION];
    char text[OFFLINE_QUEUE_MAX_TEXT];
    char client_msg_id[OFFLINE_QUEUE_MAX_MSGID];
} offline_chat_entry_t;

static offline_chat_entry_t s_oq[OFFLINE_QUEUE_CAP];
static size_t s_oq_head = 0;
static size_t s_oq_count = 0;

static bool offline_queue_enqueue(const char *scope, const char *text, const char *client_msg_id) {
    if (!text || !client_msg_id)
        return false;
    size_t tlen = strlen(text);
    if (tlen >= OFFLINE_QUEUE_MAX_TEXT) {
        /* ESP_LOGW in the real impl */
        return false;
    }
    if (strlen(client_msg_id) >= OFFLINE_QUEUE_MAX_MSGID)
        return false;
    if (scope && strlen(scope) >= OFFLINE_QUEUE_MAX_SESSION)
        return false;

    if (s_oq_count == OFFLINE_QUEUE_CAP) {
        s_oq_head = (s_oq_head + 1) % OFFLINE_QUEUE_CAP;
        s_oq_count--;
    }
    size_t slot = (s_oq_head + s_oq_count) % OFFLINE_QUEUE_CAP;
    offline_chat_entry_t *e = &s_oq[slot];
    e->scope[0] = '\0';
    if (scope) {
        strncpy(e->scope, scope, OFFLINE_QUEUE_MAX_SESSION - 1);
        e->scope[OFFLINE_QUEUE_MAX_SESSION - 1] = '\0';
    }
    strncpy(e->text, text, OFFLINE_QUEUE_MAX_TEXT - 1);
    e->text[OFFLINE_QUEUE_MAX_TEXT - 1] = '\0';
    strncpy(e->client_msg_id, client_msg_id, OFFLINE_QUEUE_MAX_MSGID - 1);
    e->client_msg_id[OFFLINE_QUEUE_MAX_MSGID - 1] = '\0';
    s_oq_count++;
    return true;
}

static size_t offline_queue_size(void) {
    return s_oq_count;
}

static bool offline_queue_peek(size_t idx, const char **out_scope, const char **out_text,
                               const char **out_client_msg_id) {
    if (idx >= s_oq_count)
        return false;
    size_t slot = (s_oq_head + idx) % OFFLINE_QUEUE_CAP;
    if (out_scope)
        *out_scope = s_oq[slot].scope;
    if (out_text)
        *out_text = s_oq[slot].text;
    if (out_client_msg_id)
        *out_client_msg_id = s_oq[slot].client_msg_id;
    return true;
}

static void offline_queue_clear(void) {
    s_oq_head = 0;
    s_oq_count = 0;
}

/* Test-only: drain by removing the head, to simulate successful flush. */
static bool offline_queue_pop_head(void) {
    if (s_oq_count == 0)
        return false;
    s_oq_head = (s_oq_head + 1) % OFFLINE_QUEUE_CAP;
    s_oq_count--;
    return true;
}

/* --------------------------------------------------------------------------
 * Tests
 * ----------------------------------------------------------------------- */

static int failed = 0;

#define EXPECT(cond, msg)                                                                                              \
    do {                                                                                                               \
        if (!(cond)) {                                                                                                 \
            fprintf(stderr, "FAIL (%s:%d): %s\n", __FILE__, __LINE__, msg);                                            \
            failed++;                                                                                                  \
        }                                                                                                              \
    } while (0)

static void test_enqueue_three_fifo_order(void) {
    offline_queue_clear();
    EXPECT(offline_queue_size() == 0, "fresh size=0");

    EXPECT(offline_queue_enqueue("s1", "hello", "m1"), "enq 1");
    EXPECT(offline_queue_enqueue("s1", "world", "m2"), "enq 2");
    EXPECT(offline_queue_enqueue("s2", "goodbye", "m3"), "enq 3");
    EXPECT(offline_queue_size() == 3, "size=3");

    const char *sk, *tx, *mi;
    EXPECT(offline_queue_peek(0, &sk, &tx, &mi), "peek 0 ok");
    EXPECT(strcmp(sk, "s1") == 0 && strcmp(tx, "hello") == 0 && strcmp(mi, "m1") == 0, "FIFO idx=0 is oldest");
    EXPECT(offline_queue_peek(1, &sk, &tx, &mi), "peek 1 ok");
    EXPECT(strcmp(sk, "s1") == 0 && strcmp(tx, "world") == 0 && strcmp(mi, "m2") == 0, "FIFO idx=1");
    EXPECT(offline_queue_peek(2, &sk, &tx, &mi), "peek 2 ok");
    EXPECT(strcmp(sk, "s2") == 0 && strcmp(tx, "goodbye") == 0 && strcmp(mi, "m3") == 0, "FIFO idx=2 is newest");
    EXPECT(!offline_queue_peek(3, &sk, &tx, &mi), "peek past end rejected");
}

static void test_overflow_drops_oldest(void) {
    offline_queue_clear();
    char text[16];
    char msgid[16];
    for (int i = 0; i < 10; i++) {
        snprintf(text, sizeof(text), "t%d", i);
        snprintf(msgid, sizeof(msgid), "m%d", i);
        EXPECT(offline_queue_enqueue("s", text, msgid), "enq");
    }
    EXPECT(offline_queue_size() == OFFLINE_QUEUE_CAP, "cap holds at 8");

    /* The oldest two (m0, m1) were dropped; queue should be m2..m9. */
    const char *sk, *tx, *mi;
    offline_queue_peek(0, &sk, &tx, &mi);
    EXPECT(strcmp(mi, "m2") == 0, "head is m2 (m0/m1 evicted)");
    offline_queue_peek(OFFLINE_QUEUE_CAP - 1, &sk, &tx, &mi);
    EXPECT(strcmp(mi, "m9") == 0, "tail is m9");
}

static void test_clear(void) {
    offline_queue_clear();
    offline_queue_enqueue("s", "a", "m1");
    offline_queue_enqueue("s", "b", "m2");
    EXPECT(offline_queue_size() == 2, "size=2 before clear");
    offline_queue_clear();
    EXPECT(offline_queue_size() == 0, "size=0 after clear");
    const char *sk, *tx, *mi;
    EXPECT(!offline_queue_peek(0, &sk, &tx, &mi), "cleared peek rejected");
}

static void test_text_cap_rejects(void) {
    offline_queue_clear();
    char big[OFFLINE_QUEUE_MAX_TEXT + 16];
    memset(big, 'x', sizeof(big) - 1);
    big[sizeof(big) - 1] = '\0';
    EXPECT(!offline_queue_enqueue("s", big, "m1"), "oversized text rejected");
    EXPECT(offline_queue_size() == 0, "size unchanged after reject");

    /* Exactly MAX_TEXT - 1 bytes fits (null terminator takes the last). */
    char atcap[OFFLINE_QUEUE_MAX_TEXT];
    memset(atcap, 'y', sizeof(atcap) - 1);
    atcap[sizeof(atcap) - 1] = '\0'; /* length = MAX_TEXT - 1 */
    EXPECT(offline_queue_enqueue("s", atcap, "m1"), "at-cap text accepted");
    EXPECT(offline_queue_size() == 1, "accepted increments size");
}

static void test_null_args_rejected(void) {
    offline_queue_clear();
    EXPECT(!offline_queue_enqueue("s", NULL, "m1"), "NULL text rejected");
    EXPECT(!offline_queue_enqueue("s", "hi", NULL), "NULL msg_id rejected");
    EXPECT(offline_queue_size() == 0, "no enqueue on NULL");
}

static void test_drain_preserves_order(void) {
    /* Simulate "flush on reconnect" as successive pop_head calls and check
     * the order we'd send to the wire. */
    offline_queue_clear();
    offline_queue_enqueue("s", "first", "m1");
    offline_queue_enqueue("s", "second", "m2");
    offline_queue_enqueue("s", "third", "m3");

    const char *sk, *tx, *mi;
    offline_queue_peek(0, &sk, &tx, &mi);
    EXPECT(strcmp(tx, "first") == 0, "drain: first");
    offline_queue_pop_head();
    offline_queue_peek(0, &sk, &tx, &mi);
    EXPECT(strcmp(tx, "second") == 0, "drain: second");
    offline_queue_pop_head();
    offline_queue_peek(0, &sk, &tx, &mi);
    EXPECT(strcmp(tx, "third") == 0, "drain: third");
    offline_queue_pop_head();
    EXPECT(offline_queue_size() == 0, "drain: empty");
}

static void test_wraparound(void) {
    /* Fill to cap, drain a couple, add more — exercises the ring wrap. */
    offline_queue_clear();
    char t[8], m[8];
    for (int i = 0; i < OFFLINE_QUEUE_CAP; i++) {
        snprintf(t, sizeof(t), "t%d", i);
        snprintf(m, sizeof(m), "m%d", i);
        offline_queue_enqueue("s", t, m);
    }
    offline_queue_pop_head(); /* drain t0 */
    offline_queue_pop_head(); /* drain t1 */
    EXPECT(offline_queue_size() == 6, "size=6 after drain 2");

    offline_queue_enqueue("s", "new1", "mn1");
    offline_queue_enqueue("s", "new2", "mn2");
    EXPECT(offline_queue_size() == 8, "refilled to cap");

    /* Head should now be t2, tail should be new2. */
    const char *sk, *tx, *mi;
    offline_queue_peek(0, &sk, &tx, &mi);
    EXPECT(strcmp(mi, "m2") == 0, "wraparound head=m2");
    offline_queue_peek(7, &sk, &tx, &mi);
    EXPECT(strcmp(mi, "mn2") == 0, "wraparound tail=mn2");
}

int main(void) {
    printf("test_enqueue_three_fifo_order...   ");
    test_enqueue_three_fifo_order();
    printf("done\n");
    printf("test_overflow_drops_oldest...      ");
    test_overflow_drops_oldest();
    printf("done\n");
    printf("test_clear...                      ");
    test_clear();
    printf("done\n");
    printf("test_text_cap_rejects...           ");
    test_text_cap_rejects();
    printf("done\n");
    printf("test_null_args_rejected...         ");
    test_null_args_rejected();
    printf("done\n");
    printf("test_drain_preserves_order...      ");
    test_drain_preserves_order();
    printf("done\n");
    printf("test_wraparound...                 ");
    test_wraparound();
    printf("done\n");

    if (failed) {
        printf("\n%d assertion(s) FAILED\n", failed);
        return 1;
    }
    printf("\nAll tests passed.\n");
    return 0;
}
